ui/src/util/convertHighlightsToReactElements.js

```diff
@@ -0,0 +1,40 @@
+import { createElement } from 'react';
+import { decode } from './htmlEntities';
+
+// Converts a string containing `<em>` tags to a list of
+// React elements without using `dangerouslySetInnerHTML`.
+// This ensures that all other HTML markup is automatically
+// escaped by React.
```

Comment:
Before, we simply rendered the text contents of a document as a text node and React would take care of escaping any HTML markup. Now, our text contents will contain some HTML markup (the `em` tags for highlighting search terms) that we do not want to be escaped. Instead of using something like `dangerouslySetInnerHTML`, this converts the text into a list of React elements (for the `em` tags) and basic text nodes. This ensures that React still escapes any HTML markup contained in the original document text and only renders the `em` tags as actual highlights.

As this is a somewhat sensitive change, please make sure to review this thoroughly, just in case I missed something. Here’s a [PDF document](https://github.com/alephdata/aleph/files/9071458/test_markup.pdf) I used during development that might be helpful.